lxd/storage/drivers/volume.go

```diff
@@ -24,3 +31,70 @@ const ContentTypeFS = ContentType("fs")
 // ContentTypeBlock indicates the volume will be a block device and its contents and we do not
 // know which filesystem(s) (if any) are in use.
 const ContentTypeBlock = ContentType("block")
+
+// Volume represents a storage volume, and provides functions to mount and unmount it.
+type Volume struct {
+	Name        string
+	Pool        string
+	Type        VolumeType
+	ContentType ContentType
+	Config      map[string]string
+	Driver      Driver
+}
+
+// NewVolume instantiates a new Volume struct.
+func NewVolume(driver Driver, poolName string, volType VolumeType, contentType ContentType, volName string, volConfig map[string]string) Volume {
+	return Volume{
+		Name:        volName,
+		Pool:        poolName,
+		Type:        volType,
+		ContentType: contentType,
+		Config:      volConfig,
+		Driver:      driver,
+	}
+}
+
+// IsSnapshot indicates if volume is a snapshot.
+func (v Volume) IsSnapshot() bool {
+	return shared.IsSnapshot(v.Name)
+}
+
+// MountPoint returns the path where the volume will be mounted.
+func (v Volume) MountPoint() string {
+	return GetVolumeMountPoint(v.Pool, v.Type, v.Name)
+}
+
+// Mount mounts the volume.
+func (v Volume) Mount(op *operations.Operation) (bool, error) {
+	return v.Driver.MountVolume(v.Type, v.Name, op)
+}
+
+// Unmount unmounts the volume.
+func (v Volume) Unmount(op *operations.Operation) (bool, error) {
+	return v.Driver.UnmountVolume(v.Type, v.Name, op)
+}
```

Comment:
For Mount/Unmount, I suspect we'll need new symbols `MountVolumeSnapshot` and `UnmountVolumeSnapshot` as snapshots do need a lot of special handling for mounting/unmounting on most drivers.

We should also document the expectation that the mountpoint of a snapshot should be treated as read-only (and we may be able to make it that way by putting a read-only bind-mount on it even for dir/cephfs).

Reply:
@stgraber makes sense, do you have any preference for where the "is snapshot" logic takes place: 

- In the driver's `MountVolume` logic (avoids needing new functions).
- In the `Volume` helper `Mount` function calling driver's `MountVolumeSnapshot`.
- In the driver's use of `Volume` helper, so adding `MountVolumeSnapshot` to Volume struct.

Reply:
I think it's reasonable for the Volume helper to do the right thing on Mount/Unmount as it's already aware of whether it's itself a snapshot or not.

Reply:
@stgraber ok cool, have added snapshot detection to volume.MountTask() and then added snapshot mount/unmount functions to driver.